File: src/main/java/by/masarnovsky/releasedemon/client/SpotifyClient.java
package by.masarnovsky.releasedemon.client;

import by.masarnovsky.releasedemon.spotify.SpotifyUserLibraryResponse;
import feign.Headers;
import org.springframework.cloud.openfeign.FeignClient;
import org.springframework.http.HttpHeaders;
import org.springframework.web.bind.annotation.RequestHeader;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;

@FeignClient(value = "spotifyclient", url = "https://api.spotify.com/v1")
public interface SpotifyClient {

    @Headers("Bearer: {token}")
    @RequestMapping(method = RequestMethod.GET, value = "/me/following?type=artist&")
    SpotifyUserLibraryResponse retrieveFollowedSpotifyArtists(@RequestParam("limit") Integer limit, @RequestParam("after") String after, @RequestHeader(HttpHeaders.AUTHORIZATION) String token);
}

File: src/main/kotlin/by/masarnovsky/releasedemon/backend/external/service/SpotifyClient.kt
package by.masarnovsky.releasedemon.backend.external.service

import by.masarnovsky.releasedemon.backend.external.dto.SpotifyUserLibraryResponse
import com.github.kittinunf.fuel.httpGet
import com.github.kittinunf.fuel.jackson.responseObject
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service

@Service
class SpotifyClient {

    @Value("\${spotify.token}")
    lateinit var token: String
    var limit = 10

    fun retrieveFollowedSpotifyArtists(after: String?): SpotifyUserLibraryResponse? {
        var spotifyUrl = "https://api.spotify.com/v1/me/following?type=artist&limit=$limit"

        if (after != null) spotifyUrl += "&after=$after"


        val (_, _, result) = spotifyUrl
            .httpGet()
            .header(mapOf("Accept" to "application/json", "Authorization" to token))
            .responseObject<SpotifyUserLibraryResponse>()

        return result.component1()
    }
}